lib/Sema/TypeCheckType.cpp

```diff
@@ -1061,6 +1061,14 @@ static Type diagnoseUnknownType(TypeResolution resolution,
         // type. Fix this by replacing 'Self' with the nominal type name.
         assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
 
+        bool insideClass = nominalDC->getSelfClassDecl() != nullptr;
+        if (!insideClass || options.isAnyExpr()) {
+          Type SelfType = nominal->getSelfInterfaceType();
```

Comment:
Sorry, can you add one more test? This should work:

```
class C {
  func f() {
    func g(_: Self) {}
  }
}
```

The lookup of `Self` is not from an expression, but it's a local declaration. You can check if `dc->isLocalContext()`. Once that's fixed I think we're good to go.

Reply:
Test added, this works now.

Reply:
... but breaks detection of:
```
  static func z(n: Self? = nil) {
```

Reply:
I've been unable to find a way of distinguishing:
```
  static func z(n: Self? = nil) {
 ```
from 
```
  func f() {
    func g(_: Self) {}
  }

Reply:
Finally settled on this code:
```C++
        bool insideClass = nominalDC->getSelfClassDecl() != nullptr;
        AbstractFunctionDecl *methodDecl = dc->getInnermostMethodContext();
        bool declaringMethod = methodDecl &&
          methodDecl->getDeclContext() == dc->getParentForLookup();

        if (!insideClass || !declaringMethod ||
            options.getBaseContext() == TypeResolverContext::ExplicitCastExpr) {
          Type SelfType = nominal->getSelfInterfaceType();
          if (insideClass)
            SelfType = DynamicSelfType::get(SelfType, ctx);
          return resolution.mapTypeIntoContext(SelfType);
        }
```

Reply:
We loose let _:Self in the process.

Reply:
What happens if you check for dc->getParent()->isLocalContext()? If its true, you allow 'Self'.